File: 2024/src/main/kotlin/solvers/Day04.kt
package org.example.solvers

import org.example.MrWolf

class Day04 : MrWolf {

    override fun solvePart1(input: String): Any {
        val grid = parse(input)
        return grid.indices.sumOf { y ->
            grid[y].indices.sumOf { x ->
                directions.count { dir -> findWord(grid, x, y, dir, "XMAS") }
            }
        }
    }

    override fun solvePart2(input: String): Any {
        val grid = parse(input)
        return grid.indices.sumOf { y ->
            grid[y].indices.count { x -> findCross(grid, x, y) }
        }
    }

    private fun parse(input: String) = input.lines().map { it.toList() }

    private fun findWord(grid: List<List<Char>>, x: Int, y: Int, dir: Direction, word: String): Boolean {
        for (index in word.indices) {
            val posX = x + dir.x * index
            val posY = y + dir.y * index
            if (!isInBounds(grid, posX, posY) || grid[posY][posX] != word[index])
                return false
        }
        return true
    }

    private fun findCross(grid: List<List<Char>>, x: Int, y: Int): Boolean {
        if (grid[y][x] != 'A') {
            return false
        }
        val upLeft = grid.getOrNull(y - 1)?.getOrNull(x - 1)
        val upRight = grid.getOrNull(y - 1)?.getOrNull(x + 1)
        val downRight = grid.getOrNull(y + 1)?.getOrNull(x + 1)
        val downLeft = grid.getOrNull(y + 1)?.getOrNull(x - 1)

        val d1 = "${upLeft}A${downRight}"
        val d2 = "${downLeft}A${upRight}"

        return (d1 == "MAS" || d1 == "SAM") && (d2 == "MAS" || d2 == "SAM")
    }

    private fun isInBounds(grid: List<List<Char>>, x: Int, y: Int): Boolean =
        y in grid.indices && x in grid[y].indices
}

private data class Direction(val x: Int, val y: Int)

private val directions = listOf(
    Direction(1, 0),    // Right
    Direction(1, 1),    // Down-Right
    Direction(0, 1),    // Down
    Direction(-1, 1),   // Down-Left
    Direction(-1, 0),   // Left
    Direction(-1, -1),  // Up-Left
    Direction(0, -1),   // Up
    Direction(1, -1),   // Up-Right
)

File: 2025/solvers/day04.ts
const solvePart1 = (input: string) => {
    const grid = parseGrid(input);
    return getAvailableRolls(grid);
};

const solvePart2 = (input: string) => {
    const grid = parseGrid(input);
    let count = 0;
    while (true) {
        const prev = count;
        count += getAvailableRolls(grid);
        if (prev === count) {
            break;
        }
    }
    return count;
};

const parseGrid = (input: string) => input.split('\n').map(row => row.split(''));

const getAvailableRolls = (grid: string[][]) => {
    const availableRolls: Pos[] = [];

    for (let y = 0; y < grid.length; y++) {
        for (let x = 0; x < grid[y].length; x++) {
            const pos: Pos = {x, y};
            if (containsRoll(pos, grid) && canBeAccessed(pos, grid)) {
                availableRolls.push(pos);
            }
        }
    }

    availableRolls.forEach(pos => grid[pos.y][pos.x] = '.');
    return availableRolls.length;
}

const canBeAccessed = (pos: Pos, grid: string[][]) =>
    directions
        .map(dir => dir(pos))
        .filter(pos => containsRoll(pos, grid))
        .length < 4;

const containsRoll = (pos: Pos, grid: string[][]) => grid[pos.y]?.[pos.x] === '@';

const up = (pos: Pos) => ({x: pos.x, y: pos.y - 1});
const down = (pos: Pos) => ({x: pos.x, y: pos.y + 1});
const left = (pos: Pos) => ({x: pos.x - 1, y: pos.y});
const right = (pos: Pos) => ({x: pos.x + 1, y: pos.y});
const upLeft = (pos: Pos) => ({x: pos.x - 1, y: pos.y - 1});
const upRight = (pos: Pos) => ({x: pos.x + 1, y: pos.y - 1});
const downLeft = (pos: Pos) => ({x: pos.x - 1, y: pos.y + 1});
const downRight = (pos: Pos) => ({x: pos.x + 1, y: pos.y + 1});

const directions = [up, down, left, right, upLeft, upRight, downLeft, downRight];

type Pos = { x: number, y: number };

export const day04 = {solvePart1, solvePart2};